Fix: Speed up bus turnaround for blackpill-f4 like other platforms

ALTracer opened the pull request:
## Detailed description

* This is not a feature, but may affect external/out-of-tree features (like direction GPIO for bus turnaround).
* The existing problem is a long function call to gpio_mode_setup() in the process of high-performance bitbanging of SWD on `blackpill-f4` family of platforms.
* This PR solves the problem by porting inline code / macros existing in other platforms.

Testing on fix/tap-timing branch with a blackpill-f411ce in max frequency (~8MHz) against a STM32MP157_CM4 shows an increase of performance of loading a 64 KiB binary into RETRAM from 156 to 180 KiB/s. Impact may be less noticeable when using Flash-based targets, or using this BMF with BMDA as proxy.

I'll mention that stlinkv3 also toggles a GPIO wired to direction pin of a bidirectional level translator in the respective macros.

## Your checklist for this pull request

* [x] I've read the [Code of Conduct](https://github.com/blackmagic-debug/blackmagic/blob/main/CODE_OF_CONDUCT.md)
* [x] I've read the [guidelines for contributing](https://github.com/blackmagic-debug/blackmagic/blob/main/CONTRIBUTING.md) to this repository
* [x] It builds for hardware native (`make PROBE_HOST=native`) _and should not affect it_
* [x] It builds as BMDA (`make PROBE_HOST=hosted`) _and should not affect it_
* [x] I've tested it to the best of my ability
* [x] My commit messages provide a useful short description of what the commits do

## Closing issues
Enhances #1688 (PR, not an Issue)

dragonmux commented on src/platforms/common/blackpill-f4/blackpill-f4.h:
```diff
@@ -115,6 +115,41 @@
 #define SWDIO_PORT TMS_PORT
 #define SWDIO_PIN  TMS_PIN
 
+#if 0
+#define TMS_SET_MODE()     gpio_mode_setup(TMS_PORT, GPIO_MODE_OUTPUT, GPIO_PUPD_NONE, TMS_PIN);
+#define SWDIO_MODE_FLOAT() gpio_mode_setup(SWDIO_PORT, GPIO_MODE_INPUT, GPIO_PUPD_NONE, SWDIO_PIN);
+
+#define SWDIO_MODE_DRIVE() gpio_mode_setup(SWDIO_PORT, GPIO_MODE_OUTPUT, GPIO_PUPD_NONE, SWDIO_PIN);
+
+#else
```
You can drop this version of the code, it's what the commit history is for 🙂

dragonmux commented on src/platforms/common/blackpill-f4/blackpill-f4.h:
```diff
@@ -115,6 +115,41 @@
+
+#else
+
+#define SWDIO_MODE_REG_MULT_PB9 (1 << (9 << 1))
+#define SWDIO_MODE_REG_MULT_PB8 (1 << (8 << 1))
```
These all need their `U` suffixes to put them in unsigned mode so we don't run the risk of an accidental conversion oops.

dragonmux commented on src/platforms/common/blackpill-f4/blackpill-f4.h:
```diff
@@ -115,6 +115,41 @@
+#define SWDIO_MODE_FLOAT()                        \
+	do {                                          \
+		uint32_t mode_reg = SWDIO_MODE_REG;       \
+		mode_reg &= ~(0x3 * SWDIO_MODE_REG_MULT); \
```
This and the '1' below in SWDIO_MODE_DRIVE() also need `U` suffixing. Might we suggest dropping the `0x` prefix here though for consistency with its counterpart below?

dragonmux reviewed (changes requested):
This looks good and provides some nice consistency with the other platforms. There are a couple of review items concerning signed-unsigned conversion to sort, but with the assumption that you've tested this with a Black Pill and it works, we've got no problems merging this once those are sorted.

dragonmux reviewed (approved):
LGTM, merging. Thank you for the contribution!

ALTracer commented:
Adjusted per review items, squashed into one; and rebased to latest main.
Tested on `PROBE_HOST=blackpill-f411ce ALTERNATIVE_PINOUT=1 BMP_BOOTLOADER=1` as well as default pinout against a gd32f103cb target (bluepillplus), everything seems to work -- JTAG/SWD, max frequency and lowered, BMF and BMDA and BMDA non-HL modes. Performance improvement estimates not taken.
